README.md
Common-Utils
============

A collection of modules which provide some useful functionality for common applications cases

Changes
Revision history for Perl extension Common::Util.

0.01  Sun Jul  1 15:20:00 2012
	- renamed Common::Utils to Common::Util

0.01  Wed Jun 20 01:03:17 2012
	- original version; created by h2xs 1.23



lib/Common/Util/File.pm
package Common::Util::File;

use 5.014002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

=head1 NAME

Common::Util::File - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Common::Util::File;

  perl -MCommon::Util::File -e 'traverse(shift || ".", { print "$_[0]\n" if $_[0] =~ /(?:\..+)?\.sw[a-z]$/ })'

=head1 DESCRIPTION

Stub documentation for Common::Util::File, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.

=cut

our %EXPORT_TAGS = ( 'all' => [ qw(
	traverse chomp read_directory
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


use Carp;

use subs qw(traverse chomp read_directory);

=head2 METHODS

=over 4

=item B<traverse>

=cut

sub traverse($\&) { # better use '&' to enable traverse(".", { do_something })
	my $dir_path = shift;
	chomp $dir_path;
	return unless -d $dir_path;
	my $handler = shift;
	read_directory($dir_path, sub {
			my $current_path = "$dir_path/$_";
			if (-f $current_path) {
				$handler->($current_path);
			} else {
				traverse($current_path, $handler);
			}
	});
}

=item B<read_directory>

=cut

sub read_directory($;&) {
	my $path = shift;
	my $handler = shift;
	my @content;
	opendir DIRECTORY, $path or croak "Can't open directory $path: $!";
	for (sort readdir DIRECTORY) {
		next if /^\.{1,2}$/;
		my $fullpath = "$path/$_";
		my $push_flag = 1;
		$push_flag = $handler->($fullpath) if defined $handler;
		push @content, $fullpath if $push_flag;
	}
	closedir DIRECTORY or carp "Can't close directory $path: $!";
	return @content;
}

=item B<chomp> Overrides the builtin C<chomp>, but as opposed to the original,
it will chomp the trailing backslash of strings which represents a valid
directory.

=cut

#NOTE, that this overridden method does NOT behave like the builtin. In fact
#it behaves like the original if a scalar or an array is given, but in case
#of omitted arguments, this method will NOT take the C<$_>-variable. So be
#careful with the import of this method!

sub chomp(;\[$@]) {
	my $string = shift;
	$string = \$_ unless defined $string;
	unless (ref $string) {
		carp "No reference given. Why?";
		$string = \$string;
	} elsif (ref $string eq 'SCALAR') {
		${$string} = $1 if ${$string} =~ /^(.+)$/;
		local $/ = '/' if -d ${$string};
		CORE::chomp(${$string});
	} else {
		for (@{$string}) {
			$_ = $1 if /^(.+)$/;
			local $/ = '/' if -d;
			CORE::chomp;
		}
	}
}

=back

=cut


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

8ware, E<lt>andydefrank@(none)E<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Common/Util/Hash.pm
package Common::Util::Hash;

use 5.014002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Common::Util::Hash ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	key_of index_of
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


use Carp;

sub key_of(\%$) {
	my %hash = %{ (shift) };
	my $value = shift;
	return undef unless defined $value;
	my @keys = grep { $hash{$_} eq $value } sort keys %hash;
	return wantarray ? @keys : $keys[0];
}


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Common::Util::Hash - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Common::Util::Hash;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Common::Util::Hash, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

8ware, E<lt>andydefrank@(none)E<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/Common/Util/Array.pm
package Common::Util::Array;

use 5.014002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Common::Util::Array ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	index_of
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

#our @EXPORT = qw(
#	
#);

our $VERSION = '0.01';


use Carp;

sub index_of(\@$) {
	my @array = @{ (shift) };
	my $value = shift;
	return undef unless defined $value;
	my @indexes = grep { $array[$_] eq $value } 0 .. $#array;
	return wantarray ? @indexes : $indexes[0];
}

# border-cases
# - negative index
# - index greater then last index +1
sub insert_at(\@$$) {
	my ($arrayref, $item, $index) = @_;
	if ($index < 1) {
		carp "Given index is negative." if $index < 0;
		unshift $arrayref, $item;
	} elsif ($index > $#{ $arrayref }) {
		carp "Given index is out of bounds." if $index > @{ $arrayref };
		push $arrayref, $item;
	} else {
		# insert_at qw(a c), 'b', 1
		@{ $arrayref } = (@{ $arrayref }[0 .. $index -1], $item,
				@{ $arrayref }[$index .. $#{ $arrayref }]);
	}
}


1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Common::Util::Array - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Common::Util::Array;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Common::Util::Array, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

8ware, E<lt>andydefrank@(none)E<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


=cut



lib/Common/Util.pm
package Common::Util;

use 5.014002;
use strict;
use warnings;

=head1 NAME

Common::Util - A collection of useful often used functionality

=head1 SYNOPSIS

  use Common::Util qw(index_of File:all Hash);

=head1 DESCRIPTION

This module provides a wrapped access to all its submodules.

=head2 EXPORT

All subroutines specified by the C<EXPORT>-array of each submodule is
imported by default. Tags can have the following structure:

  MODULE-NAME[:EXPORT-TAG]

or

  SUBROUTINE-NAME

The former variant imports all subroutines of the specified export-tag of
the module. If C<:EXPORT-TAG> is omitted the C<:all> export-tag is used.
The latter variant imports the specified subroutine of the package which
is providing that subroutine. In case of the C<:all> import-flag the
subroutines specified by the C<all>-tag of each module is imported.

=head2 PLUGIN CONVENTIONS

A plugin (prefix C<Common::Utils::>) must provide the three exporter-variables
C<@EXPORT>, C<@EXPORT_OK> and C<%EXPORT_TAGS>. All routine-names have to start
with a lower-case letter. By now, only export of subroutines is supported.

=cut

our $VERSION = '0.01';


use Carp;
use Module::Pluggable search_path => 'Common::Util', require => 1;

sub import {
	shift;
	my %export;
	my %subroutines;
	my %tags;
	for my $name (plugins()) {
		next unless $name =~ /Common::Util::(.+)/;
		no strict 'refs';
		$export{$1} = \@{"$name\::EXPORT"};
		for (@{"$name\::EXPORT_OK"}) {
			if(defined $subroutines{$_}) {
				carp "Subroutine $_ is already defined in "
						."Common::Util::$subroutines{$_}.";
			} else {
				$subroutines{$_} = $1;
			}
		}
		$tags{$1} = \%{"$name\::EXPORT_TAGS"};
	}
	use Data::Dumper;
	my %imports;
	unless (@_) {
		while (my ($module, $exports) = each %export) {
			$imports{$_} = $module for @{ $exports };
		}
	}
	for (@_) {
		given ($_) {
			when (/^([a-z]\w*)/) {
				if (defined $subroutines{$1}) {
					$imports{$1} = $subroutines{$1};
				} else {
					croak "$1 is not an exported subroutine of any module";
				}
			}
			when (':all') {
				while (my ($module, $tags) = each %tags) {
					next unless defined $tags{$module}->{all};
					$imports{$_} = $module for @{ $tags{$module}->{all} };
				}
			}
			when (':DEFAULT') {
				while (my ($module, $exports) = each %export) {
					$imports{$_} = $module for @{ $exports };
				}
			}
			when (/^([A-Z][a-z]*)(?::([a-z]+))?/) {
				my ($module, $tag) = ($1, $2 || 'all');
				if (defined $tags{$module}->{$tag}) {
					$imports{$_} = $module for @{ $tags{$module}->{$tag} };
				} else {
					croak "'$tag' is not a valid tag of Common::Util::$module";
				}
			}
			default {
				croak "$_ is not a valid export-flag";
			}
		}
	}
	my $caller = caller;
	while (my ($sub, $module) = each %imports) {
		no strict 'refs';
		*{"$caller\::$sub"} = *{"Common::Util::$module\::$sub"};
	}
}


1;
__END__

=head1 TODOs

=over 4

=item * check doubled subroutine identifier and warn for
developing purposes at beginning

=item * ensure that subroutines specified in the export-arrays/-hashes
are really defined

=item * support also export of scalars, arrays and hashes

=item * enable selection of routine-package, e.g. Array::index_of

=back

=head1 SEE ALSO

=over 4

=item Common::Util::Array

=item Common::Util::File

=item Common::Util::Hash

=back

=head1 AUTHOR

8ware, E<lt>8wared@googlemail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by 8ware

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.2 or,
at your option, any later version of Perl 5 you may have available.


=cut
